doc/developer/design/20221204_with_mutually_recursive.md

```diff
@@ -0,0 +1,183 @@
+```sql
+WITH MUTUALLY RECURSIVE
+    name_1 (col1a type1a, col1b type1b, ... ) AS ( <SELECT STATEMENT 1> ),
+    name_2 (col2a type2a, col2b type2b, ... ) AS ( <SELECT STATEMENT 2> ),
+    name_3 (col3a type3a, col3b type3b, ... ) AS ( <SELECT STATEMENT 3> ),
+    ...
+```
+This fragment binds multiple identifiers, *must* specify all column names and types, and all bound names are available in each `SELECT` statement.
+
+The value of each bound name in the following `SELECT` block (as for a standard `WITH`) is as if each were initially empty, and then all were repeatedly updated to equal their indicated definitions using the prior values of each binding.
+Mechanically, one could evaluate this as:
+```rust
+let mut prev_bindings = HashMap::new();
+let mut next_bindings = HashMap::new();
+for (name, _) in bindings.iter() {
+    prev_bindings.insert(name, empty());
+}
+
+while prev_bindings != next_bindings {
+    std::mem::swap(&mut prev_bindings, &mut next_bindings);
+    next_bindings.clear(); 
+    for (name, statement) in bindings {
+        next_bindings.insert(name, statement.eval(&prev_bindings));
+    }
+}
+```
+
+The sequence of goals have their own design considerations
+
+1.  SQL parsing
+
+    We must support the `WITH MUTUALLY RECURSIVE` fragment and capture the column names and types.
+    We are in a position to allow and capture further constraint information, such as nullability and unique keys.
+    This happens by default with `parse_columns`, though we could intentionally avoid accepting that information until we know what to do with it.
+
+2.  SQL name resolution
+
+    SQL name resolution currently introduces CTE bindings after resolving names in the CTE.
+    We would need to generalize this to introduce the names before name resolving recursive CTEs.
+
+3.  SQL planning
+
+    SQL planning currently introduces CTE bindings (types mainly) after planning the CTE.
+    We would need to generalize this to introduce the types before planning recursive CTEs. 
+
+4.  HIR generalization
+
+    HIR has a single-binding `Let` variant that cannot house mutually recursive bindings.
+    We could introduce a `LetRec` or `MultiLet` variant to support this, and even migrate other uses to it.
+
+5.  Lowering / decorrelation
+
+    It is not clear that lowering and decorrelation are complicated.
+    As existing single-binding `Let` constructs, the `value` and `body` are each prefixed with correlated columns of the outer scope, and this should continue to be the case.
+    
+6.  MIR generalization
+
+    MIR also requires a multi-binding `Let`, which could be prototyped as `LetRec` or `MultiLet`.
+
+7.  MIR optimization corrections
+
+    All MIR optimizations are at risk of being incorrect with recursive `Let` bindings.
+    Many of them propagate information about let bindings and must either a. start with conservative information (e.g. all columns required) or b. iterate to fixed point before applying their conclusions.
```

Comment:
This seems like the part with the most unknowns. Is that right, or have you done enough prototyping/thinking to derisk this part?

Reply:
I think we have a process for this, which will probably involve more than "some furious typing" like the other steps. Informally, we probably want to introduce a new `LetRec` variant that errors out with some form of `Unsupported`, and progressively implement transforms for this variant as we are certain of them. It will mean that things aren't usable until this happens, but .. seems better than starting with all the bugs and trying to guess where they came from.